sui_programmability/framework/sources/ExampleNFT.move

```diff
@@ -0,0 +1,35 @@
+// Copyright (c) 2022, Mysten Labs, Inc.
+// SPDX-License-Identifier: Apache-2.0
+
+/// A minimalist example to demonstrate how to create an NFT like object
+/// on Sui. The user should be able to use the wallet command line tool
+/// (https://docs.sui.io/build/wallet) to mint an NFT. For example,
+/// `wallet example-nft --name <Name> --description <Description> --url <URL>`
+module Sui::ExampleNFT {
+    use Sui::ID::VersionedID;
+    use Sui::TokenMetadata::{Self, TokenMetadata};
+    use Sui::Transfer;
+    use Sui::TxContext::{Self, TxContext};
+
+    /// An example NFT that can be minted by anybody
+    struct ExampleNFT has key, store {
+        id: VersionedID,
+        /// The metadata associated with this NFT
+        metadata: TokenMetadata,
+        // TODO: allow custom attributes
```

Comment:
**In general**, there are two patterns of defining custom attributes onchain:
1. as hardcoded struct fields, for example
```
struct ExampleNFT has key, store {
        id: VersionedID,
        metadata: TokenMetadata,
       level: u64,
       color: String,
}
```
The primary advantage of this approach is that it's explicit and the properties can be easily read/write by onchain programs. The downside is that it's hard to add new attribute types once the contract is deployed(before Sui support contract update).

2. as a vector of key, value pair
```
struct ExampleNFT has key, store {
        id: VersionedID,
        metadata: TokenMetadata,
       dynamic_attributes: vector<string, string>,
}
```
where attributes can be `[('level', '10'), ('color', 'blue')]`. Imagine the game company wants to add a new feature to the game by adding a `rarity` attribute to all existing NFTs, it's possible to do it in approach 2(e.g., the game company can mint a Gem that has `trait = "rarity"; value = "legendary"` and transfer to the NFT owner, the NFT owner can then invoke the `upgrade(gem, nft, ctx)` method that will add `("rarity", "legendary")` to `dynamic_attributes`). Adding new attributes in approach 1 is not possible until Sui supports contract update.

Some projects might want to use both pattern so that they can define most attributes in approach 1 but also leverage approach 2 to support future updates. The Sui Explorer should support both patterns.

For `ExampleNFT`, since it's meant to be a general NFT-like object rather than a specific item like a bored ape, I lean towards supporting custom attributes using Approach 2 where the Command Line Tool users can define there own custom attributes if they want.

Let me know if you have any thoughts cc @sblackshear @lxfind @tnowacki

Reply:
Based on Friday's discussions, I don't think we would do something exactly like (2) in Move. (though we might have some sort of immutable blob of JSON data).

If you want to make this `ExampleNFT` have custom attributes you could use generics:

```
struct ExampleNFT<Attributes: store> has key, store {
    ...
    attributes: Attributes
}
```
And then you can do 
```
struct MyRPGItemAttributes has store {
   level: u64,
   color: String,
   rarity: String,
   ...
}
...
ExampleNFT<MyRPGItemAttributes>
```

Though I'm not sure if this exactly what you want for this example

Reply:
@tnowacki I agree that approach 1 is preferred, but in the case that I mentioned above where creator(e.g., games) want to add new attributes after a contract is deployed, they have to use approach 2(at lease before Sui supports contract update).

Reply:
I see... Yeah I then we might need to define some sort of immutable JSON blob... 
If Move had maps that would be very helpful :P

Reply:
Yes, map would be really helpful here. But if I need to choose between a `vector<string, string>` and an immutable JSON string, I will prefer the former because it's easier to do on-chain operations such as add/remove/count/read for the former. And we know that a `vector` is good enough because [opensea](https://docs.opensea.io/docs/metadata-standards) is using a vector for the definition of attributes.

Reply:
You will need to do something like `vector<StringMapping>` and `struct StringMapping { key: string, value: string }` or something similarly annoying 

> And we know that a vector is good enough because [opensea](https://docs.opensea.io/docs/metadata-standards) is using a vector for the definition of attributes.

And it is "good enough", but our hope is to do better :)

Reply:
My point was that vector<(string, string)>(not sure if you can put tuples in a vec, if not then `vector<StringMapping>`) is better than an immutable JSON string. Let me know if you have better suggestions.